### PR TITLE
Add ability to hide side tab bars

### LINKED ISSUE #2031: Options to reconfigure the view to be more compact
## Summary

We are exploring ways to allow a user to collapse parts of the UI to give a more compact layout, such as hiding the side bars.

## Original issue
Is it possible to waste less of the precious vertical screen space? On a 13-inch 16:9 laptop screen there is not much vertical space to start with and in the current layout with 6 horizontal bars (OS menu bar, browser tab bar, browser adress bar, jupyterlab menu bar, jupyterlab tab bar, jupyterlab tool bar) this leaves very little space to the notebook itself.

Is there a possibility to hide the jupyterlab bars? Or to work in some single-notebook mode that would reduce the clutter?

### PR DESCRIPTION
**Pull request by fcollonval**
## References

Addresses partly #2031

hide-side-tab-bar

New view menu

image

## Code changes

- Add ability to hide side tab bars (independently of the stacked side panel - as it is possible in VS Code for example)  
The side tab bars have be called _Activity bar_ as in VS Code - I'm opened to suggestion on the naming
- Add command and shortcut to open all side panels (to cover the case where side tab bar)
- Reorganize the view menu

cc @isabela-pf @ellisonbg for the design

> UI tests are missing - but I'm waiting for feedback on the new UI before adding/updating them

## User-facing changes

New customization of the side bars
Reorganization of the _View_ menu

## Backwards-incompatible changes

N/A

### CODE REVIEW

**Comment by jtpio** [comment]
Cool stuff!

This might also relate to https://github.com/jupyterlab/jupyterlab/issues/9869 where there are some screenshots for a possible next iteration of the simple interface: https://github.com/jupyterlab/jupyterlab/issues/9869#issuecomment-786892655. In that case being able to hide the side bar tabs would be very relevant.

**Comment by krassowski** [comment]
There si a very nice mockup in https://github.com/jupyterlab/jupyterlab/issues/2031#issuecomment-295389519 though. It shows a visual indicator of the panel being collapsed and a button to uncollapse it. Sometimes users get their layout to strange state and start panicking that a part of it went missing; having an indicator of it being collapsed helps (at a cost of reducing the space gained).

**Comment by fcollonval** [comment]
> This issue says that it fixes #2031, but that issue is about hiding the vertical parts in order to maximize vertical screen space for the editors.

True I focus on the screenshot and forgot it was requesting vertical compactness too.

**Comment by fcollonval** [comment]
> having an indicator of it being collapsed helps

I'm not a big fan of small indicator. Mostly users will change the look and feel only a couple of time and then live with what they like most.

But what about a menu entry _Reset Default Layout_?

**Comment by isabela-pf** [comment]
Thanks for mentioning me! I think this makes sense as long as there is something or some very clear way to get the sidebar back. I was thinking the same thing as @krassowski mentioned above. Other than that, I think the changes to the View menu make work too.

**Review by krassowski — COMMENTED** [review]
I like it, seems like an improvement. I wonder if it would be good to expose the option to hide selected "Activity Bars" from context menus:
context-menu
The downside is that users need to know to go to View menu to show them again; if this turns out to be a problem, we could show a notification (go to view menu to restore the Left Activity Bar).

Another question is regarding the shortcuts. It is a lot of new shortcuts and there are some conflicts with system and/or browser functions; moreover we may want to use some of those shortcuts elsewhere.
- <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>F</kbd> - **?** - we may want to use that for some search related function; KDE for "Show the window in full screen mode, with no border, menubar, toolbar or statusbar"
- <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>G</kbd> - **?** - GNOME for go to previous search result; we may want to use it for search later on
- <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>R</kbd> - **?** - opens bookmark manager in browsers on KDE/GNOME or refreshes website ignoring cache
- <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>T</kbd> - **not ok** - restores previously closed tab or opens a new tab in multiple browsers
- <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>E</kbd> - fine
- <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>X</kbd> - fine

As for alternatives, we can use anything other than <kbd>Ctrl</kbd> + <kbd>Shift</kbd> +:
- <kbd>Z</kbd> - redo on KDE, GNOME, Chrome OS
- <kbd>V</kbd> - paste special on many systems (and we may want to use this)
- <kbd>Q</kbd> - can log out Chrome OS users
- <kbd>O</kbd> - opens bookmarks manager in Edge
- <kbd>N</kbd> - opens previously closed Firefox window or opens a new incognito tab on Chromium browsers
- <kbd>W</kbd> - close all windows of current application on Chrome OS, quit current application on Chrome OS

Based on https://en.wikipedia.org/wiki/Table_of_keyboard_shortcuts

**Comment by fcollonval** [comment]
Thanks for the detailed review

> The downside is that users need to know to go to View menu to show them again; if this turns out to be a problem, we could show a notification (go to view menu to restore the Left Activity Bar).

I like the idea. But as you pointed out, the users may not be able to know how to undo that action. So I propose to postpone that until we got a notification system in JLab.

FYI students at Calpoly at [working on that](https://github.com/jupytercalpoly/jupyterlab-notifications) - I'm in the loop in the hope to get it included in core after their internship.

> Another question is regarding the shortcuts.

Thanks for all the information. One question `Ctrl+Shift+F` is set to the filebrowser panel since https://github.com/jupyterlab/jupyterlab/pull/5470 - I'm unsure it is good to change it now (although this PR will make that shortcut more obvious).

What about this proposal:
* File Browser: Ctrl + Shift + F (see comment above)
* Property Inspector: Ctrl + Shift + U
* Sessions and Tabs: Ctrl + Shift + B
* Table of Contents: Ctrl + Shift + K
* Debugger Panel: Ctrl + Shift + E
* Extension Manager: Ctrl + Shift + X

**Comment by fcollonval** [comment]
@krassowski would you mind having a second look?

**Comment by krassowski** [comment]
Thanks for the ping.

- Ctrl + Shift + B -> opens bookmarks panel in Chromium browsers (not critical IMO)

The ones previously annotated with a question mark are not critical, so no strong opinions.

I agree on postponing the extra context menu, let's create an enhancement ticket for this when merging.

**Comment by fcollonval** [comment]
@krassowski Is this PR ok for you?

**Review by krassowski — APPROVED** [review]
Thanks. I wonder if it should be used in docs as "Activity Bar" (I would prefer this one) or "activity bar" (should it be added to https://jupyterlab.readthedocs.io/en/stable/developer/contributing.html#element-names as one or the other?)

**Comment by fcollonval** [comment]
I can add it to the documentation. I would also prefer "Activity Bar" and this will follow the rule of being named as it appears on the UI.

File: docs/source/user/interface.rst [hunk 1]
```diff
@@ -63,6 +63,9 @@ and a list of tabs in the main work area:
    :align: center
    :class: jp-screenshot
 
+The icons and labels column hat allows to switch between tabs is called
```
Comment (krassowski) [inline docs/source/user/interface.rst:66]:
```suggestion
The icons and labels column that allows to switch between tabs is called
```

Comment (fcollonval) [reply docs/source/user/interface.rst]: Thanks @krassowski

File: docs/source/user/interface.rst [hunk 2]
```diff
@@ -63,6 +63,9 @@ and a list of tabs in the main work area:
    :align: center
    :class: jp-screenshot
 
+The icons and labels column that allows to switch between tabs is called
+Activity Bar in JupyterLab.
```
Comment (jasongrout) [inline docs/source/user/interface.rst:67]:
```suggestion
The icons and labels column that enables switching between sidebar activities is called
Activity Bar.
```

**Comment by fcollonval** [comment]
Thanks @jasongrout for the review. Is it good to go for you?

File: docs/source/user/interface.rst (hunk 2 above)
Comment (krassowski) [reply docs/source/user/interface.rst:67]: It looks like patch from this suggestion got lost during rebase.

**Comment by LukaPitamic** [comment]
@fcollonval I just installed JLab 3.1.12, but I can't see the feature. 
image

Is there some setting to be set?

**Comment by jtpio** [comment]
@LukaPitamic it's only available in the 4.0 alpha releases (at least for now).

You can install the 4.0 pre-release if you would like to try it out:

```bash
pip install --pre -U jupyterlab
```

**Comment by LukaPitamic** [comment]
@jtpio oh, I found out about it on JLab changelog: https://jupyterlab.readthedocs.io/en/stable/getting_started/changelog.html - under _*Enhancements Made*_ list for version: 3.1.0. We're using v3.1.12 through JupyterHub - so we are slightly conservative to avoid braking the whole system. There is no other way, huh? No problem, will wait for a v4 to roll out. It's a feature we were waiting for for a long, long time (We are solving it now with CSS voodoo).
image

**Comment by jasongrout** [comment]
Looks like this issue was tagged as 3.1 between June 16 and July 12, then moved to the 4.0 milestone and merged Jul 26. It looks like it was accidentally put in the changelog for 3.1.0 in https://github.com/jupyterlab/jupyterlab/pull/10710 on Jul 27 (and https://github.com/jupyterlab/jupyterlab/pull/10709 on the 3.1.x branch). Thanks for catching this. I think it should be removed from the changelog.

**Pull request merged**